examples/weather.rs

```diff
@@ -0,0 +1,134 @@
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::net::SocketAddr;
+use std::sync::Mutex;
+use restson::{Error as RestsonError, RestPath};
+use serde::{Deserialize, Serialize};
+use jsonrpsee::{
+	ws_client::{traits::SubscriptionClient, v2::params::JsonRpcParams, WsClientBuilder},
+	ws_server::WsServer,
+	ws_server::RpcContextModule,
+};
+
+#[derive(Deserialize, Serialize, Debug, Default, PartialEq)]
+struct Weather {
+	name: String,
+	wind: Wind,
+	clouds: Clouds,
+	main: Main,
+}
+#[derive(Deserialize, Serialize, Debug, Default, PartialEq)]
+struct Clouds {
+	all: usize,
+}
+#[derive(Deserialize, Serialize, Debug, Default, PartialEq)]
+struct Main {
+	temp: f64,
+	pressure: usize,
+	humidity: usize,
+}
+#[derive(Deserialize, Serialize, Debug, Default, PartialEq)]
+struct Wind {
+	speed: f64,
+	deg: usize,
+}
+
+
+impl RestPath<&(String, String)> for Weather {
+	fn get_path(params: &(String, String)) -> Result<String, RestsonError> {
+		const API_KEY: &'static str = "f6ba475df300d5f91135550da0f4a867";
+		Ok(
+			String::from(
+				format!("data/2.5/weather?q={}&units={}&appid={}",
+					params.0,
+					params.1,
+					API_KEY,
+				)
+			)
+		)
+	}
+}
+
+/// Example of setting up a subscription that polls a remote API, in this case the api.openweathermap.org/weather, and
+/// sends the data back to the subscriber whenever the weather in London changes.
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+	env_logger::init();
+	let addr = run_server().await?;
+	let url = format!("ws://{}", addr);
+
+	let client = WsClientBuilder::default().build(&url).await?;
+
+	// Subscription to the london weather
+	let params = JsonRpcParams::Array(vec!["London,uk".into(), "metric".into()]);
+	let mut weather_sub = client.subscribe::<Weather>("weather_sub", params, "weather_unsub").await?;
+	while let Some(w) = weather_sub.next().await {
+		println!("[client] London weather: {:?}", w);
+	}
+
+	Ok(())
+}
+
+struct WeatherApiCx {
+	api_client: restson::RestClient,
+	last_weather: Weather,
+}
+
+async fn run_server() -> anyhow::Result<SocketAddr> {
+	let mut server = WsServer::new("127.0.0.1:0").await?;
+
+	let api_client = restson::RestClient::new("http://api.openweathermap.org").unwrap();
+	let last_weather = Weather::default();
+	let cx = Mutex::new(WeatherApiCx { api_client, last_weather });
```

Comment:
I suppose you could move this into closure but just an example so cool.

Reply:
That doesn't actually work, because the `restson` library is crap and requires a mutable ref to do `get()`.